commit 8aafad1c617c590596d2222717cf49a18d79f3a5
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2019-08-20T08:50:47-07:00 (7 weeks ago)
    
    TRITON-1874 add migration finalize api
    Reviewed by: Pedro Palazón Candel <pedro@joyent.com>
    Approved by: Pedro Palazón Candel <pedro@joyent.com>

diff --git a/docs/index.md b/docs/index.md
index 3517e0ed3..5d7b7a914 100644
--- a/docs/index.md
+++ b/docs/index.md
@@ -7141,7 +7141,12 @@ A migration can be set to run all of these phases in one (_automatic_ migration)
 these phases can each be run manually (_on demand_ migration).
 
 For any migration action (e.g. `begin`, `sync`, `switch` or `abort`) you can use the
-migration `watch` endpoint to show progress information for the running migration action
+migration `watch` endpoint to show progress information for the running migration action.
+
+Once the migration switch is complete, if you are happy with the new migrated
+instance then you should run the `finalize` action to remove the original
+source instance. Note that once the finalize action completes successfully, the
+migration will no longer show up in the migration list.
 
 ## Migrate  (POST /:login/machines/:id/migrate)
 
@@ -7200,13 +7205,14 @@ successful    | Migration was successfully completed.
 
 ### Migration phases
 
-The workflow stage that the migration is currently running, one of:
+The action that the migration is currently running, one of:
 
 **Phase**  | **Description**
 ---------- | ------------
 begin      | This phase starts the migration process, creates a new migration database entry and provisions the target instance.
 sync       | This phase synchronizes the zfs datasets of the source instance with the zfs datasets in the target instance (without stopping the instance).
 switch     | This phase stops the instance from running, synchronizes the zfs datasets of the source instance with the zfs datasets in the target instance, moves the NICs from the source to the target instance, moves control to the target instance and then restarts the target instance.
+finalize   | This phase is used after switch is successful - it will remove the original source instance, leaving the target instance as the one true instance.
 abort      | This phase is used when aborting a migration.
 
 ### Errors
@@ -7246,6 +7252,9 @@ MissingParameter | If `action` wasn't provided
     running: 95% starting the migrated instance
     Done - switch finished in 40.28621 seconds
 
+    $ triton instance migration finalize eaabc951
+    Done - the migration is finalized
+
     $ triton instance migration list
     SHORTID   PHASE   STATE       AGE
     eaabc951  switch  successful  26m
diff --git a/lib/migrations.js b/lib/migrations.js
index 84215a540..7138b28bd 100644
--- a/lib/migrations.js
+++ b/lib/migrations.js
@@ -19,7 +19,8 @@ var MIGRATION_ACTIONS = [
     'switch',
     'automatic',
     'abort',
-    'pause'
+    'pause',
+    'finalize'
 ];
 
 /*
@@ -242,11 +243,17 @@ function doAction(req, res, next) {
 
     var vmapiPath = util.format('/vms/%s', vmUuid);
 
-    vmapi.post(vmapiPath, opts, function doActionCb(err, out) {
+    vmapi.post(vmapiPath, opts, function doActionCb(err, out, req_, res_) {
         if (err) {
             next(err);
             return;
         }
+
+        if (params.action === 'finalize') {
+            res.send(res_.statusCode);
+            return;
+        }
+
         if (!out.job_uuid) {
             next(new restify.InternalError(
                 'Unable to execute "%s" migration action ' +
diff --git a/package.json b/package.json
index a3a578e29..7fd7dd8a6 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
     "name": "cloudapi",
     "description": "Triton CloudAPI",
-    "version": "9.8.0",
+    "version": "9.8.1",
     "author": "Joyent (joyent.com)",
     "private": true,
     "engines": {
diff --git a/test/machines/migrations.js b/test/machines/migrations.js
index 9239dd9fc..91e22a4d8 100644
--- a/test/machines/migrations.js
+++ b/test/machines/migrations.js
@@ -246,6 +246,26 @@ module.exports = function migrate(suite, client, other, machine, callback) {
         });
     });
 
+    suite.test('Finalize migration', function finalizeMigrCb(t) {
+        if (migrationFailed) {
+            t.end();
+            return;
+        }
+        var url = '/my/machines/' + machine + '/migrate';
+        client.post(url, {
+            action: 'finalize'
+        }, function doPostCb(err, req, res, body) {
+            t.ifError(err);
+            if (err) {
+                t.end();
+                return;
+            }
+
+            t.equal(res.statusCode, 200);
+            t.end();
+        });
+    });
+
     suite.test('Close migrations watch client', function closeTest(t) {
         httpClient.close();
         t.end();
